vortex-array/src/validity.rs

```diff
@@ -196,17 +196,17 @@ impl FromIterator<LogicalValidity> for Validity {
             return Self::AllInvalid;
         }
 
-        // Otherwise, map each to a bool array and concatenate them.
-        let arrays = validities
-            .iter()
-            .map(|v| {
-                v.to_present_null_buffer()
-                    .unwrap()
-                    .into_array_data()
-                    .into_array()
-            })
-            .collect::<Vec<_>>();
-        Self::Array(as_contiguous(&arrays).unwrap())
+        // Else, construct the boolean buffer
+        let mut buffer = BooleanBufferBuilder::new(validities.iter().map(|v| v.len()).sum());
+        for validity in validities {
+            let present = validity.to_present_null_buffer()
```

Comment:
I'm really not a fan of this `to_present_null_buffer` function, it can very easily lead to unnecessary allocations. Can just match on the returned optional instead.

Reply:
Hmm, so if I remove the use of `to_present_null_buffer`, I pretty much have to write

```
            let present = match validity {
                LogicalValidity::AllValid(count) => {
                    BooleanBuffer::new_set(count)
                }
                LogicalValidity::AllInvalid(count) => {
                    BooleanBuffer::new_unset(count)
                }
                LogicalValidity::Array(array) => {
                    array.into_array().flatten_bool().expect("validity must flatten to BoolArray").boolean_buffer()
                }
            };
            buffer.append_buffer(&present);
```

Which is basically identical to the implementation of `to_present_null_buffer`, except without the wrapping `NullBuffer`. I'm not sure where the unnecessary allocations happen